zkevm-circuits/src/evm_circuit/execution/calldatasize.rs

```diff
@@ -0,0 +1,184 @@
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        // Add lookup constraint in the call context for the calldatasize field.
+        let call_data_size = cb.call_context(None, CallContextFieldTag::CallDataLength);
+
+        // The calldatasize should be pushed to the top of the stack.
+        cb.stack_push(call_data_size.expr());
+
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(2.expr()),
+            program_counter: Transition::Delta(1.expr()),
+            stack_pointer: Transition::Delta((-1).expr()),
+            ..Default::default()
+        };
+
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition, None);
+
+        Self {
+            same_context,
+            call_data_size,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        _call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let call_data_size = block.rws[step.rw_indices[1]].stack_value();
+
+        self.call_data_size.assign(
+            region,
+            offset,
+            Some(Word::random_linear_combine(
+                call_data_size.to_le_bytes(),
+                block.randomness,
+            )),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::HashMap;
+
+    use bus_mapping::evm::OpcodeId;
+    use eth_types::{bytecode, Word};
+    use halo2::arithmetic::BaseExt;
+    use pairing::bn256::Fr;
+
+    use crate::evm_circuit::{
+        step::ExecutionState,
+        table::{CallContextFieldTag, RwTableTag},
+        test::{rand_bytes, run_test_circuit_incomplete_fixed_table},
+        witness::{Block, Bytecode, Call, CodeSource, ExecStep, Rw, RwMap, Transaction},
+    };
+
+    fn test_ok(call_data_size: usize, is_root: bool) {
+        let randomness = Fr::rand();
+        let bytecode = bytecode! {
+            #[start]
+            CALLDATASIZE
+            STOP
+        };
+        let bytecode = Bytecode::new(bytecode.to_vec());
+        let call_id = 1;
+        let call_data = rand_bytes(call_data_size);
+
+        let mut rw_map = HashMap::new();
+        rw_map.insert(
+            RwTableTag::CallContext,
+            vec![Rw::CallContext {
+                rw_counter: 9,
+                is_write: false,
+                call_id,
+                field_tag: CallContextFieldTag::CallDataLength,
+                value: Word::from(call_data_size),
+            }],
+        );
+        rw_map.insert(
+            RwTableTag::Stack,
+            vec![Rw::Stack {
+                rw_counter: 10,
+                is_write: true,
+                call_id,
+                stack_pointer: 1023,
+                value: Word::from(call_data_size),
```

Comment:
Following the spec, this would be RLC encoded. See https://github.com/appliedzkp/zkevm-specs/blob/feb9d0e9a42cd6ef3a62ae0d3e87cd9224100bb6/tests/evm/test_calldatasize.py#L39

Reply:
@ed255 I've used the `query_rlc` for calldatasize in the latest commit. Although this `value` field in the `Rw::Stack` is later converted into RLC-encoded value here:
https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/witness.rs#L631

Reply:
Oh ok! Got it!  Sorry for the confusion.

Reply:
Although the points you made earlier (regarding the fact that `cb.stack_pop` for the `call_data_size` value) were absolutely correct, and I've made the necessary changes for that.

Reply:
In the current state of this PR the opcode expects that in the RwTable.Stack and RwTable.CallContext.CallDataLength the values are RLC encoded.  This is because you're using `call_data_size.expr()`.  The correct implementation should be that in RwTable.Stack it's RLC encoded, but in RwTable.CallContext.CallDataLength it's the actual value.  The actual value can be obtained by using `from_bytes::expr(&call_data_size.cells)`.  Check the coinbase opcode implementation for reference.

Now, In the current state of this PR the tests are passing, so I was trying to figure out why.  And it turns out that for values between `0` and `255`, `RLC(x) == x`, which is unfortunate, because it means that when we use small values in tests we can't catch bugs when comparing RLC encoded to non-RLC encoded values.  This is because the RLC encoding does `byte0 + byte1 * rnd + byte2 * rnd**2 + ...`.  So if only byte0 is filled, the rnd is not used. 

To validate that the current implementation is not correct, please add a test case for a value of `300`, and you'll see that it fails.

Reply:
My bad, I think I should have also added a test case with a random calldatasize with upper bound on bit length. And this bug would have been caught. Thanks so much :) Will fix it today.

Reply:
@ed255 The issue should now be resolved. I've also added another constant `N_BYTES_CALLDATASIZE` which is the same as `u64` (or 8 bytes), since that's what [`go-ethereum` limits to](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L294-L297) as well.